File: sdk/php/tests/regression/RunReconnectTest.php
<?php
require 'vendor/autoload.php';

use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Generate\Futures\FuturesPublic\FuturesPublicWs;
use KuCoin\UniversalSDK\Generate\Spot\Market\GetAllSymbolsReq;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\SpotPublicWs;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use KuCoin\UniversalSDK\Model\WebSocketClientOptionBuilder;
use React\EventLoop\Loop;

function run_reconnect_test(): void
{
    $key = getenv('API_KEY') ?: '';
    $sec = getenv('API_SECRET') ?: '';
    $psw = getenv('API_PASSPHRASE') ?: '';

    $loop = Loop::get();
    $client = new DefaultClient(
        (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($sec)
            ->setPassphrase($psw)
            ->setTransportOption((new TransportOptionBuilder())->build())
            ->setWebSocketClientOption((new WebSocketClientOptionBuilder())->build())
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->build(), $loop
    );

    $rest = $client->restService();
    $wsSvc = $client->wsService();
    $symbols = array_slice(
        array_map(fn($d) => $d->symbol,
            $rest->getSpotService()->getMarketApi()
                ->getAllSymbols(GetAllSymbolsReq::builder()->setMarket('USDS')->build())->data),
        0, 50
    );

    spot_ws_example($wsSvc->newSpotPublicWs(), $symbols);
    futures_ws_example($wsSvc->newFuturesPublicWs());

    echo "Total subscribe: 53\n";
    $loop->run();
}

function spot_ws_example(SpotPublicWs $ws, array $symbols): void
{

    $promise = $ws->start();
    foreach ($symbols as $s) {
        $promise = $promise->then(function () use ($ws, $s) {
            return $ws->trade([$s], static function () {
            });
        });
    }

    $promise->then(function () use ($ws) {
        return $ws->ticker(['BTC-USDT', 'ETH-USDT'], static function () {
        });
    })->then(function () use ($ws) {
        echo "Spot subscribe [OK]\n";
    })->catch(function ($error) use ($ws) {
        echo "Spot subscribe [FAIL]" . $error->getMessage() . "\n";
        exit - 1;
    });
}

function futures_ws_example(FuturesPublicWs $ws): void
{
    $ws->start()->then(function () use ($ws) {
        return $ws->tickerV2('XBTUSDTM', static function () {
        });
    })->then(function () use ($ws) {
        return $ws->tickerV1('XBTUSDTM', static function () {
        });
    })->then(function () use ($ws) {
        echo "Futures subscribe [OK]\n";
    }, function ($error) use ($ws) {
        echo "Futures subscribe [Error]" . $error->getMessage() . "\n";
    });
}

run_reconnect_test();

File: sdk/java/example/src/main/java/com/kucoin/test/regression/RunReconnectTest.java
package com.kucoin.test.regression;

import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.api.KucoinRestService;
import com.kucoin.universal.sdk.api.KucoinWSService;
import com.kucoin.universal.sdk.generate.futures.futurespublic.FuturesPublicWs;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsData;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsReq;
import com.kucoin.universal.sdk.generate.spot.spotpublic.SpotPublicWs;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import com.kucoin.universal.sdk.model.WebSocketClientOption;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.stream.Collectors;
import lombok.extern.slf4j.Slf4j;

/** Runs a WebSocket reconnect test, subscribing to spot and futures market data. */
@Slf4j
public class RunReconnectTest {

  public static void runReconnectTest() {
    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    ClientOption clientOption =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .transportOption(TransportOption.builder().build())
            .websocketClientOption(WebSocketClientOption.defaults())
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .build();

    KucoinClient client = new DefaultKucoinClient(clientOption);
    KucoinRestService rest = client.getRestService();
    KucoinWSService wsSvc = client.getWsService();

    List<String> symbols =
        rest
            .getSpotService()
            .getMarketApi()
            .getAllSymbols(GetAllSymbolsReq.builder().market("USDS").build())
            .getData()
            .stream()
            .map(GetAllSymbolsData::getSymbol)
            .limit(50)
            .collect(Collectors.toList());

    spotWsExample(wsSvc.newSpotPublicWS(), symbols);
    futuresWsExample(wsSvc.newFuturesPublicWS());

    log.info("Total subscribe: 53");

    try {
      CountDownLatch latch = new CountDownLatch(1);
      latch.await();
    } catch (InterruptedException e) {
      log.error("Main thread interrupted", e);
      Thread.currentThread().interrupt();
    }
  }

  private static void noop(String __, String ___, Object ____) {}

  public static void spotWsExample(SpotPublicWs ws, List<String> symbols) {
    ws.start();
    for (String symbol : symbols) {
      ws.trade(new String[] {symbol}, RunReconnectTest::noop);
    }
    ws.ticker(new String[] {"BTC-USDT", "ETH-USDT"}, RunReconnectTest::noop);
  }

  public static void futuresWsExample(FuturesPublicWs ws) {
    ws.start();
    ws.tickerV2("XBTUSDTM", RunReconnectTest::noop);
    ws.tickerV1("XBTUSDTM", RunReconnectTest::noop);
    log.info("Futures subscribe [OK]");
  }

  public static void main(String[] args) {
    runReconnectTest();
  }
}
